[Bugfix] Fix vllm_flash_attn rotary import

jeejeelee opened the pull request:
Using qwen2-vl can reproduce this issue.
FIX https://github.com/vllm-project/vllm/issues/17263

Co-authored-by: Aaron Pham <contact@aarnphm.xyz>

jeejeelee committed:
Done

Signed-off-by: Jee Jee Li <pandaleefree@gmail.com>

aarnphm commented on setup.py:
```diff
@@ -280,6 +281,26 @@ def run(self):
             self.copy_file(file, dst_file)
 
+        # copy vllm_flash_attn/layers dir
+        src_layers = os.path.join(self.build_lib, "vllm", "vllm_flash_attn",
+                                  "layers")
+        dst_layers = os.path.join("vllm", "vllm_flash_attn", "layers")
+        if os.path.exists(src_layers):
+            print(f"Copying directory {src_layers} to {dst_layers}")
+            if os.path.exists(dst_layers):
+                shutil.rmtree(dst_layers)
+            shutil.copytree(src_layers, dst_layers)
+
+        # copy vllm_flash_attn/ops dir
+        src_ops = os.path.join(self.build_lib, "vllm", "vllm_flash_attn",
+                               "ops")
+        dst_ops = os.path.join("vllm", "vllm_flash_attn", "ops")
+        if os.path.exists(src_ops):
+            print(f"Copying directory {src_ops} to {dst_ops}")
+            if os.path.exists(dst_ops):
+                shutil.rmtree(dst_ops)
+            shutil.copytree(src_ops, dst_ops)
```
You should use `self.copy_tree` instead

```suggestion
				for folder in ("layers", "ops"):
            src = os.path.join(self.build_lib, "vllm", "vllm_flash_attn", folder)
            out = os.path.join("vllm", "vllm_flash_attn", folder)
            self.copy_tree(src, out)
```

aarnphm replied:
You can use this patch here

```patch
diff --git a/setup.py b/setup.py
index 671ec7739..f5b3f5165 100755
--- a/setup.py
+++ b/setup.py
@@ -283,6 +283,13 @@ class cmake_build_ext(build_ext):
             dst_file = os.path.join("vllm/vllm_flash_attn", basename)
             print(f"Copying {file} to {dst_file}")
             self.copy_file(file, dst_file)
+        # We will need to copy these folders to use its ops for qwen-vl
+        for folder in ("layers", "ops"):
+            src = os.path.join(self.build_lib, 'vllm', 'vllm_flash_attn',
+                               folder)
+            out = os.path.join('vllm', 'vllm_flash_attn', folder)
+            print(f"Copying {folder} from vllm/vllm_flash_attn")
+            self.copy_tree(src, out)
 
 
 class repackage_wheel(build_ext):
@@ -402,6 +409,32 @@ class repackage_wheel(build_ext):
 
                 package_data[package_name].append(file_name)
 
+            # Extract and include the layers and ops folders for qwen-vl
+            folders_to_copy = {"layers", "ops"}
+            for folder in folders_to_copy:
+                folder_path = f"vllm/vllm_flash_attn/{folder}"
+                folder_files = [
+                    f for f in wheel.filelist
+                    if f.filename.startswith(folder_path)
+                ]
+
+                if folder_files:
+                    print(f"Include {folder} folder from vllm/vllm_flash_attn")
+                    for file in folder_files:
+                        wheel.extract(file)
+
+                        # Add the file to package_data if it's not a Python file
+                        rel_path = file.filename.split("/")
+                        # vllm/vllm_flash_attn/folder/file
+                        if len(rel_path) >= 4:
+                            package_name = "vllm.vllm_flash_attn." + folder
+                            file_name = rel_path[-1]
+
+                            if not file_name.endswith(".py"):
+                                if package_name not in package_data:
+                                    package_data[package_name] = []
+                                package_data[package_name].append(file_name)
+
 
 def _is_hpu() -> bool:
     # if VLLM_TARGET_DEVICE env var was set explicitly, skip HPU autodetection
```

jeejeelee replied:
Thank you

aarnphm reviewed (changes requested):
One slight change

jeejeelee committed:
Modify code

Signed-off-by: Jee Jee Li <pandaleefree@gmail.com>

aarnphm reviewed (approved):
Thanks. LGTM

Can you also add a `Co-authored-by: Aaron Pham <contact@aarnphm.xyz>` to the description.

aarnphm commented:
Hmm, @jeejeelee there is a issue with building on CPU here?

aarnphm commented:
Ah we need to gated the copy ovee in `_is_cuda()` only here.

LucasWilkinson commented on setup.py:
```diff
@@ -280,6 +280,14 @@ def run(self):
             self.copy_file(file, dst_file)
 
+        # copy these folders to use the vllm_flash_attn rotary_kernel.
+        for folder in ("layers", "ops"):
```
this seems unnecessary, why not must do:
```
        files = glob.glob(
            os.path.join(self.build_lib, "vllm", "vllm_flash_attn", "**",  "*.py"))
```
on line 275-276 instead?

aarnphm replied:
this will only grab all of the python files, then we still need to check if the folder exists or not, then copy over here.

jeejeelee replied:
Would this conflict with https://github.com/vllm-project/vllm/pull/17260?

LucasWilkinson commented on setup.py:
```diff
@@ -400,6 +408,31 @@ def run(self) -> None:
 
+            # Extract and include the layers and ops of rotary embedding.
+            folders_to_copy = {"layers", "ops"}
+            for folder in folders_to_copy:
```
This also seems unnecessarily messing / complicated, I think we can just use `glob.translate` (https://docs.python.org/3/library/glob.html#glob.translate) and grab the file list like use the same glob pattern mentioned above

aarnphm replied:
i agree, overlook on my end.

LucasWilkinson reviewed (changes requested):
Can you please add a description of the bug to the PR comment

jeejeelee commented:
> Thanks. LGTM
> 
> Can you also add a `Co-authored-by: Aaron Pham <contact@aarnphm.xyz>` to the description.

Done

jeejeelee commented:
> Ah we need to gated the copy ovee in `_is_cuda()` only here.
> 
> ```diff
> 27fdbeea7 - chore: only gated in CUDA (HEAD -> fix-flash-att-rotray)
> 
> Signed-off-by: Aaron Pham <contact@aarnphm.xyz>
> (G) Aaron Pham <contact@aarnphm.xyz> (Sun Apr 27 14:57:44 2025 +0000)
> 
> 
> diff --git a/setup.py b/setup.py
> index 34be6a641..8e18c8215 100755
> --- a/setup.py
> +++ b/setup.py
> @@ -280,13 +280,14 @@ class cmake_build_ext(build_ext):
>              print(f"Copying {file} to {dst_file}")
>              self.copy_file(file, dst_file)
>  
> -        # copy these folders to use the vllm_flash_attn rotary_kernel.
> -        for folder in ("layers", "ops"):
> -            src = os.path.join(self.build_lib, "vllm", "vllm_flash_attn",
> -                               folder)
> -            out = os.path.join("vllm", "vllm_flash_attn", folder)
> -            print(f"Copying {folder} from vllm/vllm_flash_attn")
> -            self.copy_tree(src, out)
> +        if _is_cuda():
> +            # copy these folders to use the vllm_flash_attn rotary_kernel.
> +            for folder in ("layers", "ops"):
> +                src = os.path.join(self.build_lib, "vllm", "vllm_flash_attn",
> +                                   folder)
> +                out = os.path.join("vllm", "vllm_flash_attn", folder)
> +                print(f"Copying {folder} from vllm/vllm_flash_attn")
> +                self.copy_tree(src, out)
>  
>  
>  class repackage_wheel(build_ext):
> @@ -409,29 +410,34 @@ class repackage_wheel(build_ext):
>                  package_data[package_name].append(file_name)
>  
>              # Extract and include the layers and ops of rotary embedding.
> -            folders_to_copy = {"layers", "ops"}
> -            for folder in folders_to_copy:
> -                folder_path = f"vllm/vllm_flash_attn/{folder}"
> -                folder_files = [
> -                    f for f in wheel.filelist
> -                    if f.filename.startswith(folder_path)
> -                ]
> -
> -                if folder_files:
> -                    print(f"Include {folder} folder from vllm/vllm_flash_attn")
> -                    for file in folder_files:
> -                        wheel.extract(file)
> -
> -                        # Add the file to package_data if it's not a Python file
> -                        rel_path = file.filename.split("/")
> -                        # vllm/vllm_flash_attn/folder/file
> -                        if len(rel_path) >= 4:
> -                            package_name = "vllm.vllm_flash_attn." + folder
> -                            file_name = rel_path[-1]
> -                            if not file_name.endswith(".py"):
> -                                if package_name not in package_data:
> -                                    package_data[package_name] = []
> -                                package_data[package_name].append(file_name)
> +            if _is_cuda():
> +                folders_to_copy = {"layers", "ops"}
> +                for folder in folders_to_copy:
> +                    folder_path = f"vllm/vllm_flash_attn/{folder}"
> +                    folder_files = [
> +                        f for f in wheel.filelist
> +                        if f.filename.startswith(folder_path)
> +                    ]
> +
> +                    if folder_files:
> +                        print(
> +                            f"Include {folder} folder from vllm/vllm_flash_attn"
> +                        )
> +                        for file in folder_files:
> +                            wheel.extract(file)
> +
> +                            # Add the file to package_data if
> +                            # it's not a Python file
> +                            rel_path = file.filename.split("/")
> +                            # vllm/vllm_flash_attn/folder/file
> +                            if len(rel_path) >= 4:
> +                                package_name = "vllm.vllm_flash_attn." + folder
> +                                file_name = rel_path[-1]
> +                                if not file_name.endswith(".py"):
> +                                    if package_name not in package_data:
> +                                        package_data[package_name] = []
> +                                    package_data[package_name].append(
> +                                        file_name)
>  
>  
>  def _is_hpu() -> bool:
> ```

IIUC, no need to add gate, these changes are just a supplement to `vllm_flash_attn` copy

aarnphm commented:
This is superceded by #17267
